Try: Register a fake wc-admin homepage to for users without permission

joshuatf opened the pull request:
Fixes (possibly) #7061 

Note that this branch is attempting to shift responsibilities to the client for capability checks, however it is impossible to fully remove some of the server-side checks and some of the workarounds in this PR feel less than ideal.

I will open a 2nd PR soon to explore shifting some responsibility back to the server and doing this the more "WordPress way."

Users without the `manage_woocommerce` permission were not able to see any wc-admin registered pages.  This PR:
* Registers wc-admin with the cap check of `read` to allow everyone to access `page=wc-admin` pages.
* Removes said menu item if the user cannot manage WooCommerce

### Screenshots

<img width="190" alt="Screen Shot 2021-07-01 at 4 17 12 PM" src="https://user-images.githubusercontent.com/10561050/124184899-384e6000-da88-11eb-9fc5-95ba8d0a27d0.png">


### Detailed test instructions:

1. Create a user with a capability of `read` and `view_woommerce_reports` and `view_admin_dashboard`
2. Log in with that user
3. Check that you can access Analytics pages, but the WooCommerce menu item and homepage are not available
4. Add the capability `manage_woocommerce`
5. Make sure that the a "WooCommerce" menu item is shown that shows the wc-admin homepage
6. Make sure no regressions have occurred with default admin accounts

joshuatf committed:
Register a fake wc-admin homepage to for users without permission

joshuatf committed:
Update removal cap to manage_woocommerce

joshuatf committed:
Fix WooCommerce menu item position and icon

joshuatf committed:
Fix up admin user checks

joshuatf commented on src/Features/Homescreen.php:
```diff
@@ -74,18 +75,63 @@ public function add_user_data_fields( $user_data_fields ) {
 	public function register_page() {
+		// Register a top-level item for users who cannot view the core WooCommerce menu.
+		if ( ! $this->is_admin_user() ) {
+			add_filter( 'woocommerce_prevent_admin_access', '__return_false' );
```
This will need to be replaced, but was a short-term fix.  Core allows the following capabilities without redirection:

* `edit_posts`
* `manage_woocommerce`
* `view_admin_dashboard`

The `view_admin_dashboard` is not a traditional capability and was created for this purpose, however it's also not a default available capability and may create confusion for store owners.

Alternatively (and more realistically) we may need to capture all the registered pages in wc-admin and check against the user capabilities to see if any match before turning this off.  Will update accordingly if we decide to go the direction of this PR.

louwie17 replied:
I take it that we only need to register this for the redirection? Given we remove it from the menu.

Also what makes this menu item show up in the `global $menu` when the one registered below doesn't show up, is it just the use of `order`?

joshuatf replied:
>I take it that we only need to register this for the redirection? Given we remove it from the menu.

The `woocommerce_prevent_admin_access` filter is for redirection, but this menu item is registered to allow users to access `wc-admin` pages otherwise they may not have the capability to access a `wc-admin` page even though the page has a different capability.

WordPress only cares about the `?page=wc-admin` portion and ignores `path`.

>Also what makes this menu item show up in the global $menu when the one registered below doesn't show up, is it just the use of order?

The one below has a parent of `woocommerce` which requires certain capabilities the user may or may not have.  This item is registered as a top-level item.

louwie17 replied:
Gotcha, thanks for the clarification @joshuatf, my WP PHP knowledge is a bit limited, this helps a lot.

louwie17 commented on src/Features/Homescreen.php:
```diff
@@ -74,18 +75,63 @@ public function add_user_data_fields( $user_data_fields ) {
+					'path'       => self::MENU_SLUG,
+					'position'   => 10,
+					'capability' => 'read',
+					'icon'       => $woocommerce_icon,
```
We should be able to remove the icon usage, as it is removed from the menu right?

joshuatf replied:
I think you're right.  Can't see a scenario where we'll need it at the moment.

louwie17 commented on src/Features/Homescreen.php:
```diff
@@ -74,18 +75,63 @@ public function add_user_data_fields( $user_data_fields ) {
+		}
+
+		foreach ( $menu as $key => $menu_item ) {
+			if ( 'wc-admin' !== $menu_item[2] || 'read' !== $menu_item[1] ) {
```
Maybe use `self::MENU_SLUG` here instead of `wc-admin`?

joshuatf replied:
Good call.

louwie17 reviewed (commented):
I left a couple small comments/questions, but it tested well!

> I will open a 2nd PR soon to explore shifting some responsibility back to the server and doing this the more "WordPress way."

I would be curious to see this approach, before we potentially go this route.

joshuatf committed:
Handle PR feedback

joshuatf committed:
Only disable redirection on wc-admin pages

joshuatf commented:
Thanks for the review @louwie17.  Addressed your feedback and I updated the `woocommerce_prevent_admin_access` to only disable on wc-admin pages.  I think this will require a follow-up in the future if we merge this in, but this may be a viable short-term fix.

louwie17 commented on src/Features/Homescreen.php:
```diff
@@ -74,18 +75,68 @@ public function add_user_data_fields( $user_data_fields ) {
+				'woocommerce_prevent_admin_access',
+				function( $bool ) {
+					// phpcs:disable WordPress.Security.NonceVerification.Recommended
+					if ( isset( $_GET['page'] ) && 'wc-admin' === $_GET['page'] ) {
```
Do we also want to add `wc-reports` to this, as acceptable? And probably use `self::MENU_SLUG` here again for the `$_GET['page']`.

louwie17 reviewed (commented):
Thanks for the updates @joshuatf, things tested well! I am good to go ahead with this as it does fix the issue at hand, aside from the small suggestion below. We should continue to iterate on this as it is certainly not perfect. The points you mentioned in your last P2 comment -> p90Yrv-2k6-p2#comment-5909 sum it up well.

> Thanks for the review @louwie17. Addressed your feedback and I updated the woocommerce_prevent_admin_access to only disable on wc-admin pages. I think this will require a follow-up in the future if we merge this in, but this may be a viable short-term fix.

Thanks for narrowing down the `woocommerce_prevent_admin_access` filter, I think this makes sense, the only caveat is that now the `Dashboard` and `Profile` menu items link back to the users `My account` (non admin) page, but I think that's ok.
I just left one suggestion on adding `wc-reports` page as well -> https://github.com/woocommerce/woocommerce-admin/pull/7285/files#r666262127 (unless we are depreciating `wc-reports`?)

joshuatf commented:
>Thanks for narrowing down the woocommerce_prevent_admin_access filter, I think this makes sense, the only caveat is that now the Dashboard and Profile menu items link back to the users My account (non admin) page, but I think that's ok.

Good point.  I think this is a good argument for not messing with that filter.  Maybe we should stick to using the `view_admin_dashboard` capability.  Even though it's a bit difficult to find, there is prior work in place to make it function as is.

joshuatf committed:
Remove woocommerce_prevent_admin_access filter

joshuatf commented:
@louwie17 I removed the filter and updated the test instructions; specifically this bit:

>1. Create a user with a capability of `read` and `view_woommerce_reports` and `view_admin_dashboard`

This works out of the box with the WooCommerce filter and allows the Profile/Dashboard to work as well.  

The only downside is that one more capability is needed and the capability is a bit obscure (I can't seem to find any official docs for it, although many random tutorials do seem to mention it).

joshuatf commented:
Hey @louwie17 this is ready for another look when you have time.

louwie17 reviewed (approved):
> I removed the filter and updated the test instructions

Nice! Thanks for the updates @joshuatf, things continue to test well and code looks great!

The E2E tests job failed, I am guessing that is related to something else, or might just need a re-run?

mattsherman commented:
@joshuatf @louwie17 Shall we keep this PR open? Just checking, since it has sat in "needs review" for a week with no activity (we've had a lot of other issues, so perhaps expected!).

joshuatf commented:
@mattsherman I think that was just related to failing e2e tests.  Passing now, so I'll merge this one.

louwie17 commented:
Yeah this was good to merge aside from the E2E test failure.

joshuatf committed:
Add changelog entry
